core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala

```diff
@@ -149,29 +162,70 @@ class BlockManagerMasterEndpoint(
+      new mutable.HashMap[BlockManagerId, mutable.HashSet[RDDBlockId]]
+
     blocks.foreach { blockId =>
-      val bms: mutable.HashSet[BlockManagerId] = blockLocations.get(blockId)
-      bms.foreach(bm => blockManagerInfo.get(bm).foreach(_.removeBlock(blockId)))
-      blockLocations.remove(blockId)
+      val bms: mutable.HashSet[BlockManagerId] = blockLocations.remove(blockId)
+
+      val (bmIdExtShuffle, bmIdExecutor) = bms.partition(_.port == externalShuffleServicePort)
+      if (bmIdExecutor.isEmpty && bmIdExtShuffle.nonEmpty) {
+        // when original executor is already removed use the shuffle service to remove the blocks
+        bmIdExtShuffle.foreach { bmIdForShuffleService =>
+          val blockIdsToDel = blocksToDeleteByShuffleService.getOrElseUpdate(bmIdForShuffleService,
+            new mutable.HashSet[RDDBlockId]())
+          blockIdsToDel += blockId
+          blockStatusByShuffleService.get(bmIdForShuffleService).foreach { blockStatus =>
+            blockStatus.remove(blockId)
+          }
+        }
+      } else {
+        bmIdExecutor.foreach { bm =>
+          blockManagerInfo.get(bm).foreach { bmInfo =>
+            bmInfo.removeBlock(blockId)
+          }
+        }
+      }
     }
+    val removeRddFromExecutorsFutures = blockManagerInfo.values.map { bmInfo =>
+      askRemoveRddFromExecutor(removeMsg, bmInfo)
+    }.toSeq
 
-    // Ask the slaves to remove the RDD, and put the result in a sequence of Futures.
-    // The dispatcher is used as an implicit argument into the Future sequence construction.
-    val removeMsg = RemoveRdd(rddId)
-
-    val futures = blockManagerInfo.values.map { bm =>
-      bm.slaveEndpoint.ask[Int](removeMsg).recover {
-        case e: IOException =>
-          logWarning(s"Error trying to remove RDD $rddId from block manager ${bm.blockManagerId}",
-            e)
-          0 // zero blocks were removed
+    val removeRddBlockViaExtShuffleServiceFutures = externalShuffleClient.map { shuffleClient =>
+      blocksToDeleteByShuffleService.map { case (bmId, blockIds) =>
+        Future[Int] {
+          val numRemovedBlocks = shuffleClient.removeBlocks(
+            bmId.host,
+            bmId.port,
+            bmId.executorId,
+            blockIds.map(_.toString).toArray)
+          numRemovedBlocks.get(defaultRpcTimeout.duration.toSeconds, TimeUnit.SECONDS)
+        }
       }
-    }.toSeq
+    }.getOrElse(Seq.empty)
 
-    Future.sequence(futures)
+    Future.sequence(removeRddFromExecutorsFutures ++ removeRddBlockViaExtShuffleServiceFutures)
+  }
+
+  /**
+   * Ask the slaves to remove the RDD.
+   */
+  private def askRemoveRddFromExecutor(
+      removeMsg: RemoveRdd,
+      bmInfo: BlockManagerInfo): Future[Int] = {
+    bmInfo.slaveEndpoint.ask[Int](removeMsg).recover {
+      case e: IOException =>
+        logWarning(s"Error trying to remove RDD ${removeMsg.rddId} " +
+          s"from block manager ${bmInfo.blockManagerId}", e)
+        0 // zero blocks were removed
```

Comment:
In certain situations (e.g. executor death) it would make sense to retry this through the shuffle service. But I'm not sure how to safely detect those situations (or whether it makes sense to always retry through the shuffle service).

Reply:
I will think about that and If nothings comes to my mind I can add a TODO or open a Jira issue about this question. Is that Ok?

Reply:
Jira issue is created: https://issues.apache.org/jira/browse/SPARK-27819